File: v1/LLD/Design-Patterns/Iterator/code/Iterator.java
// Iterator Pattern — a Playlist exposes a PlaylistIterator instead of its internal array.
// Compile: javac Iterator.java
// Run:     java Iterator

import java.util.ArrayList;
import java.util.List;

interface SongIterator {
    boolean hasNext();
    String next();
}

class Playlist {
    private List<String> songs = new ArrayList<>();

    void addSong(String song) {
        songs.add(song);
    }

    SongIterator createIterator() {
        return new PlaylistIterator(songs);
    }
}

class PlaylistIterator implements SongIterator {
    private List<String> songs;
    private int position = 0;

    PlaylistIterator(List<String> songs) {
        this.songs = songs;
    }

    public boolean hasNext() {
        return position < songs.size();
    }

    public String next() {
        String song = songs.get(position);
        position++;
        return song;
    }
}

public class Iterator {
    public static void main(String[] args) {
        Playlist playlist = new Playlist();
        playlist.addSong("Clair de Lune");
        playlist.addSong("Gymnopedie No.1");
        playlist.addSong("River Flows in You");

        SongIterator iterator = playlist.createIterator();

        System.out.println("Playing playlist:");
        while (iterator.hasNext()) {
            System.out.println("Now playing: " + iterator.next());
        }
    }
}

File: v1/LLD/Design-Patterns/Iterator/code/iterator.py
"""
Iterator Pattern — a Playlist exposes a PlaylistIterator instead of its internal list.
Run: python iterator.py
"""


class Playlist:
    def __init__(self):
        self.songs = []

    def add_song(self, song):
        self.songs.append(song)

    def create_iterator(self):
        return PlaylistIterator(self.songs)


class PlaylistIterator:
    def __init__(self, songs):
        self.songs = songs
        self.position = 0

    def has_next(self):
        return self.position < len(self.songs)

    def next(self):
        song = self.songs[self.position]
        self.position += 1
        return song


if __name__ == "__main__":
    playlist = Playlist()
    playlist.add_song("Clair de Lune")
    playlist.add_song("Gymnopedie No.1")
    playlist.add_song("River Flows in You")

    iterator = playlist.create_iterator()

    print("Playing playlist:")
    while iterator.has_next():
        print(f"Now playing: {iterator.next()}")
